relay-server/src/extractors/shared_payload.rs

```
@@ -46,6 +48,29 @@ impl SharedPayload {
         self.inner.unread_data(chunk);
         self.done.store(false, Ordering::Relaxed);
     }
+
+    /// Stream a chunk of the raw request body.
+    ///
+    /// When the request body has been exhausted, this will return None.
+    pub async fn chunk(&mut self) -> Result<Option<Bytes>, PayloadError> {
+        if self.done.load(Ordering::Relaxed) {
+            return Ok(None);
+        }
+
+        // Resolve the next chunk from the actix payload stream (futures 0.1 compatibility).
+        let result = (&mut self.inner).compat().try_next().await;
+
+        if let Ok(None) = result {
+            self.done.store(true, Ordering::Relaxed);
+        }
+
+        result
+    }
+
+    /// Consume the remaining response body.
+    pub async fn consume(mut self) {
+        while let Ok(Some(_)) = self.chunk().await {}
```

If `self.chunk()` returns an error, this function will silently fail. Is that the same behavior as we had before?

---

Yes, we want to consume until there's an error and then bail.